### DISCUSSION TITLE
Community Guidelines around Tech-/Preview Features, token-exchange

### DISCUSSION THREAD

**Discussion by cwpenhale** [keycloak/keycloak#12794] [Q&A]
Hello Keycloak Community!

The guidelines around tech-preview / preview features are clear [in the documentation](https://www.keycloak.org/server/features):
"Preview features are disabled by default and are not recommended for use in production. These features may change or be removed at a future release."

What is unclear to me is what the best way of interacting with the development community on questions like:

1.  When is a feature expected to graduate to supported?
2. When is a feature at risk of being dropped?
3. Other than not being considered production ready, what does preview mean? Can we plan to use it in the future if it's working for us in a lower environment?

We're trying to make a decision on whether or not work with the token-exchange feature. We see active development and bugs being logged in the GitHub issue tracker, which gives us some confidence, but we'd like to know the best channel for navigating to production with these features, to plan our releases.

With that in mind, is there an open issue tracking token-exchange as a feature? If so, is there a Keycloak release coming up where it's expected to be in the supported phase? Finally, how does the community feel today about using token-exchange?

Thanks!
Connor

**Comment by cgeorgilakis** [comment]
For us Token Exchange for OIDC clients as described in [OAuth 2.0 Token Exchange documentation](https://datatracker.ietf.org/doc/html/rfc8693) is an important feature and we have enabled it in all our Keycloak instances.
The main problem is that Keycloak does not support scope parameter as mentioned in the [OAuth 2.0 Token Exchange documentation](https://datatracker.ietf.org/doc/html/rfc8693). We have changed code in order to support scope parameter only for OIDC clients. However, we do not know how scope parameter is used in other cases of Token Exchange. We do not use them.

We have maid a [PR](https://github.com/keycloak/keycloak/pull/10904). We are using now Keycloak 16.1 and we investigate migrating to 18. After version 16.1, Keycloak has extra changes in token exchange validation not yet investigated. Taking into account this limitation, I propose to enable and use Token Exchange.
